crates/re_query_cache/src/cache.rs

```diff
@@ -0,0 +1,364 @@
+use std::{
+    collections::{BTreeMap, VecDeque},
+    sync::Arc,
+};
+
+use ahash::HashMap;
+use itertools::Either;
+use once_cell::sync::Lazy;
+use parking_lot::RwLock;
+use paste::paste;
+use seq_macro::seq;
+
+use re_data_store::LatestAtQuery;
+use re_log_types::{EntityPath, RowId, StoreId, TimeInt, Timeline};
+use re_query::ArchetypeView;
+use re_types_core::{components::InstanceKey, Archetype, ArchetypeName, Component, ComponentName};
+
+use crate::{ErasedFlatVecDeque, FlatVecDeque};
+
+// ---
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum AnyQuery {
+    LatestAt(LatestAtQuery),
+    // TODO(cmc): range queries support.
+}
+
+impl From<LatestAtQuery> for AnyQuery {
+    fn from(query: LatestAtQuery) -> Self {
+        Self::LatestAt(query)
+    }
+}
+
+// ---
+
+// TODO(cmc): Centralize and harmonize all caches (query, jpeg, mesh).
+static CACHES: Lazy<Caches> = Lazy::new(Caches::default);
+
+/// Maintains the top-level cache mappings.
+//
+// TODO(cmc): Store subscriber and cache invalidation.
+// TODO(#4730): SizeBytes support + size stats + mem panel
+// TODO(cmc): timeless caching support
+#[derive(Default)]
+pub struct Caches {
+    latest_at: RwLock<HashMap<CacheKey, Arc<RwLock<LatestAtCache>>>>,
+}
+
+impl Caches {
+    /// Clears all caches.
+    //
+    // TODO(#4731): expose palette command.
+    #[inline]
+    pub fn clear() {
+        let Caches { latest_at } = &*CACHES;
+        latest_at.write().clear();
+    }
+
+    /// Gives write access to the appropriate `LatestAtCache` according to the specified
+    /// query parameters.
+    #[inline]
+    pub fn with_latest_at<A, F, R>(
+        store_id: StoreId,
+        entity_path: EntityPath,
+        query: &LatestAtQuery,
+        mut f: F,
+    ) -> R
+    where
+        A: Archetype,
+        F: FnMut(&mut LatestAtCache) -> R,
+    {
+        let key = CacheKey::new(store_id, entity_path, query.timeline, A::name());
+
+        // We want to make sure we release the lock on the top-level cache map ASAP.
+        let cache = {
+            let mut caches = CACHES.latest_at.write();
+            let latest_at_cache = caches.entry(key).or_default();
+            Arc::clone(latest_at_cache)
+        };
+
+        let mut cache = cache.write();
+        f(&mut cache)
+    }
+}
+
+/// Uniquely identifies cached query results in the [`Caches`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct CacheKey {
+    /// Which [`re_data_store::DataStore`] is the query targeting?
+    pub store_id: StoreId,
+
+    /// Which [`EntityPath`] is the query targeting?
+    pub entity_path: EntityPath,
+
+    /// Which [`Timeline`] is the query targeting?
+    pub timeline: Timeline,
+
+    /// Which [`Archetype`] are we querying for?
+    ///
+    /// This is very important because of our data model: we not only query for components, but we
+    /// query for components from a specific point-of-view (the so-called primary component).
```

Comment:
We should probably pick one terminology and stick with it ("primary" or "pov")

Reply:
We've discussed this during standpoint: we'll continue down that road for now and see where it leads... :grimacing: